### DISCUSSION TITLE
Help for dissertation in BSc CS

### DISCUSSION THREAD

**Discussion by akonst08** [ultralytics/ultralytics#24313] [Q&A] [answered]
Hello All, 

I hope you are well. I would like some help on an issue I have on my thesis and I should mention that my timeline is very short.
Now about my topic-concern:
I have a YOLOv11 detector trained on 8 hysteroscopic lesion classes (medical), but I now received about 20–30 videos that contain endometritis (lesion) and I do not have frame-level annotations or bounding boxes. I only know at video level that endometritis is present, and I have no clinician support to identify where it appears (specific time of the video). I need the fastest practical pipeline to mine high-probability candidate frames, generate pseudo-labels, and train an additional detection class without retraining everything from scratch. My current concern is that the 8-class detector may not detect anything in these videos, so candidate mining should not depend on the existing detector. Please propose a step-by-step, time-efficient, code-oriented workflow using anomaly ranking, temporal consistency, SAM-assisted region proposals, and iterative pseudo-label filtering.
My dissertation won't be published, however is an important matter that would lead to my graduation. I spent many hours, running experiments that required several hours and I had no help at all, however due to time limitation I am a bit stressed. 

I would appreciate any help and advices and thanks for your time reading this!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @akonst08, thank you for your interest in Ultralytics 🚀, and thanks for sharing the context around your dissertation. This is an automated response to help get things moving quickly; an Ultralytics engineer will also assist you soon 😊

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

For your case, since this looks like a training/workflow question under time pressure, it would help a lot if you could share:
-Your current `ultralytics` version
-A short description of your existing 8-class detector setup
-A few sample frames or screenshots from the videos
-Any current candidate-mining or pseudo-labeling scripts you have already tried
-Training logs, errors, or output examples from your latest experiments
-If applicable, a small [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the current blocker

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

-**Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-**Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-**Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-**Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply, marked as answer]
Given your timeline, the fastest practical route is weakly supervised mining rather than direct 9-class retraining: extract frames at 1-2 FPS, drop blurry and near-duplicate frames, rank candidates detector-independently with image-embedding outlier scores, keep only temporally consistent runs of 3-5 consecutive high-score frames, use SAM on those runs for region proposals, manually verify only the top few hundred boxes to build a small seed set, then fine-tune from your existing `best.pt` on a mixed 9-class dataset plus a small replay subset of the original 8 classes to reduce forgetting; after that, pseudo-label the remaining videos with high `conf` and `save_txt=True`, keep boxes only if they are stable across adjacent frames and pass simple size/aspect checks, retrain once more, and stop after 1-2 iterations—the closest Ultralytics reference is this [semi-supervised learning guide](https://www.ultralytics.com/glossary/semi-supervised-learning), and if helpful I can sketch the exact folder layout and minimal `train()`/`predict()` script next.

**Reply by akonst08** [reply]
Therefore, should I use something like ResNet after extracting the frames to rank them and then proceed to SAM? I would be looking at this from tomorrow and I would come back. 

Thanks for your reply and help!

**Reply by glenn-jocher** [reply]
Yes—that’s a sensible order: extract frames first, rank them with a pretrained image encoder (`ResNet` is fine for speed, though `CLIP`/`DINO` embeddings usually separate anomalies better), keep only short temporally consistent high-score runs, then send those frames to SAM for proposals and use the verified boxes for pseudo-labeling; the closest Ultralytics references are the [semi-supervised learning guide](https://www.ultralytics.com/glossary/semi-supervised-learning) and the [active learning guide](https://www.ultralytics.com/glossary/active-learning), and when you come back I can turn that into a minimal script.

**Reply by akonst08** [reply]
Instead of 20-30 videos, I was given 8 videos and a medical report indicating 19 frames that show lesion (However, they are not labeled with the exact region that the lesion appears.). I built an initial pipeline as an automated annotator, that extracts the frames from the pdf, sends them into DINO (shouldn't have done that in this anyway) and then I am using SAM to estimate the regions. Yet, I saw that SAM draws boundaries of regions very very badly. I mean (imagine the image outside of hysteroscopic view which is a total black that considers those as different regions... and the whole of hysteroscopic as a one.) I did all these to at least have a base seed of more official labeled frames before proceeding with the rest. But still I am unsure about it, how would I ensure that are more valid and formal those labels? Or should I stop thinking of that and proceed with what I got ? 

Thanks again in for your time.

**Reply by glenn-jocher** [reply]
I wouldn’t treat those raw SAM masks as labels yet; for hysteroscopy, first crop or mask the circular field of view so the black border is excluded, then use SAM only for candidate regions and convert the reviewed outputs into coarse detection boxes, because a small clean seed set from those 19 positive frames is much safer than many noisy labels—the key part is annotation quality control, as covered in the [data collection and annotation guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/), and then you can fine-tune from your current weights and pseudo-label the remaining videos.
